### PR TITLE
feat: Add config option for max concurrent requests

### PR DESCRIPTION
**Pull request by loewenheim**
This adds a symbolicator config option `max_concurrent_requests: Option<usize>` that limits the number of requests symbolicator processes concurrently; any incoming requests beyond this number will be rejected. When this happens, `create_symbolication_request` returns a `MaxRequestsError` that is later converted to a `503` (`Internal Server Error`) status code by actix_web.

The option's default value is currently `None`, which corresponds to the status quo of never rejecting requests. The default will change once we have gathered some production data about request numbers.

(Also there's a bonus unrelated clippy fix)

### CODE REVIEW

File: crates/symbolicator/src/endpoints/applecrashreport.rs [hunk 1]
```diff
@@ -46,8 +46,13 @@ async fn handle_apple_crash_report_request(
     let report = report.ok_or_else(|| error::ErrorBadRequest("missing apple crash report"))?;
 
     let symbolication = state.symbolication();
-    let request_id =
-        symbolication.process_apple_crash_report(params.scope, report, sources, options);
+    let request_id = symbolication
+        .process_apple_crash_report(params.scope, report, sources, options)
+        .ok_or_else(|| {
+            error::ErrorServiceUnavailable(
+                "Symbolicator is currently processing the maximum number of requests.",
```
Comment (relaxolotl) [inline crates/symbolicator/src/endpoints/applecrashreport.rs:53]: I assume this is the error message you're referring to that's a bit of a WIP: I'm thinking we could maybe say something like "Cannot process this request: symbolicator has reached the maximum number of concurrent requests ({what the maximum is})".

Is this string being returned as-is to sentry? Who exactly sees this error string?

Comment (loewenheim) [reply crates/symbolicator/src/endpoints/applecrashreport.rs:53]: The WIP message I was referring to was literally "TODO", but I'm not super happy with this one either. Now that I look at them I see that other error messages in these files are very terse, and IIUC nobody should ever actually see these.

Comment (relaxolotl) [reply crates/symbolicator/src/endpoints/applecrashreport.rs]: 👍 the new message looks a lot better and is in line with the existing ones in the project

**Comment by relaxolotl** [comment]
It would be great if it was possible to add in at least two basic unit tests verifying that the limit does indeed get enforced and that setting no limit also works, but I'm unsure of the amount of work it would take to get this running. 

@Swatinem would you happen to know if this would require building quite a bit of new infra and/or needs an unnecessarily large amount of effort relative to the actual size of the change itself?

**Comment by Swatinem** [comment]
I don’t think its particularly difficult:

submit symbolication requests that are accessing a non-responding symbol server. see that the request that goes over the quota is rejected immediately.
Its just a question of how to not lock up when it comes to async rust.

File: crates/symbolicator/src/endpoints/applecrashreport.rs [hunk 2]
```diff
@@ -46,8 +46,11 @@ async fn handle_apple_crash_report_request(
     let report = report.ok_or_else(|| error::ErrorBadRequest("missing apple crash report"))?;
 
     let symbolication = state.symbolication();
-    let request_id =
-        symbolication.process_apple_crash_report(params.scope, report, sources, options);
+    let request_id = symbolication
+        .process_apple_crash_report(params.scope, report, sources, options)
+        .ok_or_else(|| {
+            error::ErrorServiceUnavailable("maximum number of concurrent requests reached")
+        })?;
```
Comment (Swatinem) [inline crates/symbolicator/src/endpoints/applecrashreport.rs:53]: I’m wondering why we `ok_or_else` here, and why not just straight up return the error from the appropriate functions.

File: crates/symbolicator/src/services/symbolication.rs [hunk 3]
```diff
@@ -420,10 +427,18 @@ impl SymbolicationActor {
         // Assume that there are no UUID4 collisions in practice.
         let requests = self.requests.clone();
 
-        if let Ok(num_requests) = requests.lock().len().try_into() {
+        let num_requests = requests.lock().len();
+        if let Ok(num_requests) = num_requests.try_into() {
             metric!(gauge("requests.in_flight") = num_requests);
         }
```
Comment (Swatinem) [inline crates/symbolicator/src/services/symbolication.rs:433]: this `try_into` is converting a usize to u64, right? We have a bunch of `as u64` in various places already, so might as well use that.

**Review by Swatinem — COMMENTED** [review]
lgtm, though I would rather return a proper `MaxCurrencyReached` error, which has a sensible `Display` impl.

**Review by relaxolotl — APPROVED** [review]
thanks for adding in the tests!

File: crates/symbolicator/src/services/symbolication.rs [hunk 4]
```diff
@@ -2795,4 +2802,34 @@ mod tests {
...
+        };
+        let service = Service::create(config).unwrap();
+        let symbolication = service.symbolication();
+        let symbol_server = test::FailingSymbolServer::new();
+
+        test::spawn_compat(move || async move {
+            // Make three requests that never get resolved. Since the server is configured to only accept a maximum of
+            // two concurrent requests, the first two should succeed and the third one should fail.
+            let request = get_symbolication_request(vec![symbol_server.pending_source.clone()]);
+            symbolication.symbolicate_stacktraces(request).unwrap();
```
Comment (Swatinem) [inline crates/symbolicator/src/services/symbolication.rs:2825]: `assert!(....is_ok());`

reads a bit nicer than the unwrap. Esp. below `.is_err()` vs `unwrap_err()`.

**Review by Swatinem — APPROVED** [review]

File: crates/symbolicator/src/services/symbolication.rs (hunk 4 above)
Comment (loewenheim) [reply crates/symbolicator/src/services/symbolication.rs]: You're totally right, fixed.

**Pull request merged**